doc/developer/design/20220203_connectors.md

```diff
@@ -0,0 +1,211 @@
+
+While technically supporting ALTER does not have to require changing the catalog representation of sources, the utility of supporting ALTER without CASCADE is very small and supporting CASCADE requires that relations are discoverable.
+
+## Staged Implementation
+Since there is significant user experience value in the base feature without ALTER, not being able to ALTER is not a regression since sources are already immutable, and implementation is significantly less complex if connectors are only used during initial planning of a `CREATE SOURCE` statement I propose we break things into several distinct stages:
+
+1. Connectors as syntactic sugar during `CREATE SOURCE`
+2. Refactor Sources in the catalog such that connectors are referenced by relation (either implicitly or explicitly)
+    - unblock `ALTER ... CASCADE` behavior in the planner 
+    - unlocks other desirable features as well
+3. Determine exact semantics around `ALTER` and running sources once Platform has decoupled Sources into the STORAGE layer
+    - Upsert command?
+    - Consistency semantics between layers
+      - If eventual for running instance, how to convey progress to user   
+4. Implement `CASCADE` in planner/ADAPTER and source/STORAGE pieces according to the new semantics
+
+## Reference
+
+### Kafka connector
+```ebnf
+create_connector_kafka ::=
+  'CREATE' 'CONNECTOR' ('IF NOT EXISTS')? connector_name
+  'FOR' 'KAFKA BROKER' host ('USING AUTH' (ssl_security_protocol_options | sasl_plain_protocol_options | sasl_gssapi_protocol_option) )?
+ssl_security_protocol_options ::=
+  '(' 'security_protocol' '=' ('ssl' | 'sasl_ssl') ','
+  'ssl_key_location' '=' ssl_key_location ','
+  'ssl_certificate_location' '=' ssl_certificate_location ','?
+  ('ssl_ca_location' '=' ssl_ca_location ','
+  'ssl_key_password' '=' ssl_key_password)? ')'
+sasl_plain_protocol_options ::=
+  '(' 'security_protocol' '=' ('sasl_plaintext' | 'plaintext') ','
+  'sasl_mechanisms' '=' ('PLAIN' | 'SCRAM-SHA-256' | 'SCRAM-SHA-512') ','
+  'sasl_username' '=' sasl_username ','
+  ( ('sasl_password' '=' sasl_password) | ('sasl_password_env' '=' sasl_password_env) ) ')'
+sasl_gssapi_protocol_option ::=
+  '(' 'security_protocol' '=' ('sasl_plaintext' | 'plaintext') ','
+  'sasl_mechanisms' '=' 'GSSAPI' ','
+  'sasl_kerberos_keytab' '=' sasl_kerberos_keytab ','
+  'sasl_kerberos_kinit_cmd' '=' sasl_kerberos_kinit_cmd ','
+  'sasl_kerberos_min_time_before_relogin' '=' sasl_kerberos_min_time_before_relogin ','
+  'sasl_kerberos_principal' '=' sasl_kerberos_principal ','
+  'sasl_kerberos_service_name' '=' sasl_kerberos_service_name ')'
```

Comment:
So, a few thoughts on this syntax!

* The EBNFs as written constrain the order in which parameters must be presented. Was that the intent?

* We're trying to draw a distinction in the SQL parser between `WITH` options, which are arbitrary key–value pairs from the perspective of the grammar, and proper syntax, which uses `SQL KEYWORDS` instead of `underscored_identifiers`. For example, the new `CREATE CLUSTER` syntax looks like this:
  ```sql
   CREATE CLUSTER foo [<cluster option> [, <cluster option> ...]]

   cluster_option ::=
       VIRTUAL [[=] <value>]
       REMOTE '(' <value> [, <value> ] ')
       SIZE [[=] <value>]
       INTROSPECTION GRANULARITY <value>
       INTROSPECTION DEBUGGING <value>
   ```

  Also, it's not entirely clear to me what's a string literal in the grammar and what's not, but as a rule the SQL parser is not allowed to match on the contents of string literals. If it's something the SQL parser is meant to parse, it should be either a token or a keyword.

  * Can you remind me of our goals here? I'm sorry, my memory of our meeting is fuzzy. I though I remembered @elindsey saying the librdkafka options were complicated enough that it would be easier just to keep them as `WITH` options.

As I see it written out, I'm a bit nervous about the potential for scope creep if we try to do the syntax redesign and the `CONNECTOR` split at the same time. I'm wondering if we should make our lives easier by doing a very _direct_ extraction of the connector bit from `CREATE SOURCE` while leaving the syntax alone, and then coming back to the question of the syntax redesign. If nothing else, might be a good way to structure the work. We can leave `CREATE CONNECTOR` as experimental until we decide whether we want to punt the syntax redesign or not.

Reply:
My understanding from our conversation had been that we did not want to bring `WITH` options over to connectors and instead we should limit them to just the necessary syntax which is what I was trying to do here. I did my best to keep the changes to syntax extremely minimal, mostly it just lifts the specific required parameters out of `WITH` options. I will do another pass over the grammar to make it conform with convention on capitalization to make clear what are keywords and what aren't.

Reply:
I think there's a small set of Kafka connector options that are dictated by the auth mechanisms supported in both Apache and Confluent Kafka that are easily turned into keywords. The following isn't quite perfect but I think conveys the scope of what I'm suggesting.
```ebnf
CREATE CONNECTOR connector_name
FOR KAFKA 
  BROKER <value>
  SECURITY <security_options>

security_options ::=
  SSL
    KEY_FILE [=] <value>
    CERTIFICATE_FILE [=] <value>
    KEY_PASSWORD [[=] <value>]
  SASL
    MECHANISMS <value>
    USERNAME [=] <value>
    PASSOWRD [[=] <value>]
    PASSWORD_ENV [[=] <value>]
  GSSAPI
    KEYTAB [=] <value>
    KINIT_CMD [=] <value>
    RELOGIN_TIME [=] <value>
    PRINCIPAL [=] <value>
    SERVICE_NAME [=] <value>
```

That said now that I am understanding the rest of your comments properly I agree that it makes sense to do the syntax changes in a later phase.

Reply:
Yeah, that looks pretty solid! We may want to introduce some well placed parens to make the association clear:

```sql
security_options ::=
  SSL (
    KEY_FILE [=] <value>
    CERTIFICATE_FILE [=] <value>
    KEY_PASSWORD [[=] <value>]
  )
  SASL (
    MECHANISMS <value>
    USERNAME [=] <value>
    PASSOWRD [[=] <value>]
    PASSWORD_ENV [[=] <value>]
  )
  GSSAPI (
    KEYTAB [=] <value>
    KINIT_CMD [=] <value>
    RELOGIN_TIME [=] <value>
    PRINCIPAL [=] <value>
    SERVICE_NAME [=] <value>
  )
```

But yeah, doing this in two phases seems like the right way to structure the design and coding work!